### ISSUE TITLE
Printing unicode text on LED Panel

### ISSUE THREAD

**Issue by pgbhat** [hzeller/rpi-rgb-led-matrix#711]
Please guide me how to display an Indic language characters on LED Panel - e.g., 0C85 to 0C8A.

**Comment by hzeller** [comment]
It should be possible to use the [otf2bdf](https://github.com/jirutka/otf2bdf)-tool to convert any TrueType font to BDF (There might be an issue though if these are outside the 16-Bit range of the old unicode standard, AFAIK the BDF format does not allow full 32-bit unicode. But it looks like it should for for characters such as U+0C85).

You can simply install otf2bdf on your debian machine with `sudo apt-get install otf2bdf`.

Now, use the TrueType font that contains the characters and run otf2bdf; in the following example, we are reading the true-type font `path-to-ttf-file.ttf`, and converting it to a 28pt size bitmap BDF font, stored in `/tmp/output.bdf`
```
   otf2bdf -v -o /tmp/output.bdf -r 72 -p 28 path-to-ttf-file.ttf
```

You then can use the BDF font, using the regular font API provided by the LED matrix (you can try it first with the `examples-api-use/text-example` provided in the distribution).

**Comment by hzeller** [comment]
The API itself [accepts text in UTF-8 encoding](https://github.com/hzeller/rpi-rgb-led-matrix/blob/master/include/graphics.h#L83).

**Comment by hzeller** [comment]
Did you get it to work ?

**Comment by pgbhat** [comment]
Thank you so much, Henner. Your input helped me to find a solution. 

Also, it is very kind of you enquiring if it worked. 

 

Regards,

PG Bhat

**Comment by hzeller** [comment]
Cool, I can close this issue then.

**Comment by arahasya** [comment]
Hi Henner,
After a lot of research on this topic
The above method of font convert won't work for languages like Devnagiri, Hebrew, Urdu 
Basically languages with complex ligatures and substitutions. 
The only way is to add support for OpenType text shaping libraries which handle these tasks like Harfbuzz.
Or a way around is use Qt for render text but it will require dependency on Qt libraries.
Do you have any ideas on how to achieve this for this library?

**Comment by hzeller** [comment]
Yes, unfortunately, the old BDF fonts (from the 80ies) can only deal with Plane 0 unicode (u0000-uFFFF). And while Devnagiri, Hebrew and Urdu are in that plane, the current simplistic implementation assumes that the language is a LTR language and that each glyph is on its own without diacritic modifiers, which doesn't work for many languages. RTL could probably be implemented fairly straight forward, but more complex text shaping can not really be achieved with the current approach.

I am happy to accept a pull request if you'd like to add better text rendering, but keep the dependencies to a minimium, so Qt would be a bit heavy, more something along FreeType and HarfBuzz. I'd say start out with a separate implementation, that isn't tied to the current simplistic [`Font`](https://github.com/hzeller/rpi-rgb-led-matrix/blob/master/include/graphics.h#L28) but rather something that is useful in your context. We then later can see how we can consolidate simple BDF rendering and fully featured text shaping.

As for implementation: In general, this library provides a fairly simple [`Canvas`](https://github.com/hzeller/rpi-rgb-led-matrix/blob/master/include/canvas.h#L32) interface, so your text rendering essentially just needs to be able to call `SetPixel()`. In fact, the current [`DrawText()`](https://github.com/hzeller/rpi-rgb-led-matrix/blob/master/include/graphics.h#L121) is implemented entirely independent of the LED matrix, its only assumption is that there is a Canvas to draw to.

**Comment by arahasya** [comment]
So I have found a working example. But I'm not sure how to integrate this. Can you have a look?

https://github.com/tangrams/harfbuzz-example

**Comment by arahasya** [comment]
Hi,
So I tried rendering with Qt. Text is getting rendered of all fonts of all languages but there are some problems. There is some offset in x,y positions in the actual rendered text on matrix. It is different for different fonts. The spacing between the letters is uneven. This is because in monochrome mode sub-pixel rendering is not getting done properly. I tried all other settings but getting no better result.
As Qt also used FreeType, harfbuzz they will have the same problem. Do you have any solution in mind to this problem?
Only bitmap fonts work properly for such low resolutions. 

Comparison :-
1. Qt Full colour on black window 

![WhatsApp Image 2020-08-27 at 16 43 03](https://user-images.githubusercontent.com/59508143/91752101-75855f80-ebe3-11ea-82fc-5a07feaba43b.jpeg)

2. Qt monochrome on black window.

![WhatsApp Image 2020-08-27 at 16 43 03](https://user-images.githubusercontent.com/59508143/91752149-859d3f00-ebe3-11ea-909c-8b4ea9534959.jpeg)

Notice the spacing and quality

**Comment by arahasya** [comment]
@hzeller?

**Comment by hzeller** [comment]
Yes, for low resolution you need to look into bitmap fonts. TrueType fonts are not really meant for that, but if you use them make sure that all hints are used (I suppose TrueType has similar low-resolution hints as Type1 fonts have).

**Comment by marcmerlin** [comment]
With apologies to everyone, this project is doing a one time cleanup of 
all the bugs that have not been addressed and closed in the last 10 years.

If you filed an issue to ask for help, please do not re-open this one or a new
bug, but go here instead: https://rpi-rgb-led-matrix.discourse.group/

Please be mindful that this project does not have any full time maintainers
or anyone paid. You should not expect personal help for your own projects or
answers to programming questions but you can ask for help with panels that
aren't working, setup questions, or anything that may be a bug here:
https://rpi-rgb-led-matrix.discourse.group/
If you need programming help and are willing to pay someone for it, please
state so in your message, and someone may be interested.

If you did report an actual bug that is still not fixed today, please accept
our apologies for this being mass closed, you can absolutely send a followup
and re-open the issue, or open a new up to date one as applicable.
Note that we very much welcome help, so if you can help fix your bug, you will
be a superstar :)

If your bug was linked to a pull request that was not resolved, please update
the pull request and the issue, and we will do our best to review them. We do
thank you for your undertanding that we have very limited free time, so any
extra work you do along with your bug report or PR request will be very much
appreciated and improve the odds that they will be looked at.

If you want to help contribute as a python or C# or any other language bindings,
you are also a superstar, please state so in the issue and we'll do our best to
give you the access and reviews you need.

For anything related to PWM panel support, please make sure you read README and
then refer to https://github.com/hzeller/rpi-rgb-led-matrix/issues/466 . 
Please do not comment if you're haven't read everything and/or if you're just
asking when your panel is going to be supported. Those bugs are if you're 
able to contribute to the effort.

For rPI5: https://github.com/hzeller/rpi-rgb-led-matrix/issues/1603

For anything python, please link your bug to: https://github.com/hzeller/rpi-rgb-led-matrix/issues/1828
For anything C#, please link your bug to: https://github.com/hzeller/rpi-rgb-led-matrix/issues/1828
